packages/image_picker/android/src/main/java/io/flutter/plugins/imagepicker/ThumbnailCreator.java

```diff
@@ -0,0 +1,92 @@
+import android.os.AsyncTask;
+import android.provider.MediaStore;
+import androidx.core.util.Consumer;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+
+public class ThumbnailCreator {
+
+  private final ImageResizer imageResizer;
+  private final File externalFilesDirectory;
+
+  public ThumbnailCreator(ImageResizer imageResizer, File externalFilesDirectory) {
+    this.imageResizer = imageResizer;
+    this.externalFilesDirectory = externalFilesDirectory;
+  }
+
+  /** returns the path for the thumbnail image. */
+  public String generateImageThumbnail(String originalImagePath, Double width, Double height) {
+    return imageResizer.resizeImageIfNeeded(originalImagePath, width, height);
+  }
+
+  /** returns the path for the thumbnail image. */
+  public void generateImageThumbnailAsync(
+      final String originalImagePath,
+      final Double width,
+      final Double height,
+      Consumer<String> callback) {
+    BitmapAsyncGenerator generator =
+        new BitmapAsyncGenerator(
+            callback,
+            new Supplier<String>() {
+              @Override
+              public String get() {
+                return generateImageThumbnail(originalImagePath, width, height);
+              }
+            });
+    generator.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+  }
+
+  /** returns the path for the thumbnail video. */
+  public void generateVideoThumbnailAsync(
+      final String originalVideoPath,
+      final Double width,
+      final Double height,
+      Consumer<String> callback) {
+    BitmapAsyncGenerator generator =
+        new BitmapAsyncGenerator(
+            callback,
+            new Supplier<String>() {
+              @Override
+              public String get() {
+                return generateVideoThumbnail(originalVideoPath, width, height);
+              }
+            });
+    generator.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+  }
+
+  public String generateVideoThumbnail(String originalVideoPath, Double width, Double height) {
+    try {
+      Bitmap bitmap =
+          ThumbnailUtils.createVideoThumbnail(
+              originalVideoPath, MediaStore.Images.Thumbnails.MINI_KIND);
+
+      Bitmap scaledBmp =
+          Bitmap.createScaledBitmap(bitmap, width.intValue(), height.intValue(), false);
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      scaledBmp.compress(Bitmap.CompressFormat.JPEG, 100, outputStream);
+
+      String[] pathParts = originalVideoPath.split("/");
+      String imageName = pathParts[pathParts.length - 1];
+
+      File imageFile = new File(externalFilesDirectory, "/thumbnail" + imageName);
+      FileOutputStream fileOutput = new FileOutputStream(imageFile);
+      fileOutput.write(outputStream.toByteArray());
+      fileOutput.close();
+
+      return imageResizer.resizeImageIfNeeded(originalVideoPath, width, height);
+    } catch (IOException e) {
```

Comment:
Catching and rethrowing the exception will make failures here hard to debug since it's obfuscating the original type of problem. Can we handle the I/O exceptions instead?